### PR TITLE
Using Write + Seek types for output

### PR DESCRIPTION
**Pull request by cessen**
This pull request enables any `Write + Seek` type to be used as the writer for OpenEXR output.

### CODE REVIEW

File: src/output/stream_writer.rs
Comment (Ralith) [inline src/output/stream_writer.rs]: This type (and by extension, the entire module) looks like dead code to me.

File: src/stream_io.rs
Comment (Ralith) [inline src/stream_io.rs]: This seems like a lot of jumping through ugly hoops just to use OpenEXR's built-in file I/O functionality, [which doesn't appear to do anything special](https://github.com/openexr/openexr/blob/develop/OpenEXR/IlmImf/ImfStdIO.cpp). Letting people use Rust `File` objects through the generic interface instead would make for prettier and more obvious types and reduce the size of the library.

If this approach is taken, the CEXR code for constructing from a file path should probably be removed as dead code, and `from_writer` should be renamed to `new`.

File: src/output/stream_writer.rs
Comment (Ralith) [inline src/output/stream_writer.rs]: Unless a generic parameter is explicitly qualified as `?Sized`, I don't think this is a concern.

File: tests/memory_io.rs
Comment (Ralith) [inline tests/memory_io.rs:32]: Since this gets overwritten on line 60, it should be zero-initialized for clarity and to ensure we don't report a false positive if the read operation fails to have any effects.

File: tests/memory_io.rs
Comment (Ralith) [inline tests/memory_io.rs]: I don't think this inner block is necessary. You're declaring `mut fb: FrameBuffer` regardless, so no name pollution is being avoided, and the inlined version is shorter. Same goes for the similar block in the writing section.

File: src/stream_io.rs
Comment (Ralith) [inline src/stream_io.rs]: We should never throw away error data, due to the added difficulty this can entail in debugging. [`io::Error::raw_os_error`](https://doc.rust-lang.org/std/io/struct.Error.html#method.raw_os_error) could be used here to provide a value we can easily return into C++ and pass back out again.

File: src/stream_io.rs
Comment (Ralith) [inline src/stream_io.rs]: See error handling comment on `write_stream`.

File: openexr-sys/c_wrapper/rust_ostream.cpp
Comment (Ralith) [inline openexr-sys/c_wrapper/rust_ostream.cpp]: The cause of the error should not be thrown away. If the rust function is modified to return an OS error, then a exception type containing the OS error code could be thrown here instead, which would allow for better messages and could eventually be leveraged to convert failures back into a Rust `io::Error` downstream. I believe the correct exception type to use here to get helpful messages with no extra effort would be a [`std::system_error`](http://en.cppreference.com/w/cpp/error/system_error/system_error) using [`std::system_category`](http://en.cppreference.com/w/cpp/error/system_category).

File: openexr-sys/c_wrapper/rust_ostream.cpp
Comment (Ralith) [inline openexr-sys/c_wrapper/rust_ostream.cpp]: See error handling comment on `write`.

**Review by Ralith — CHANGES_REQUESTED** [review]
Looks pretty good! I spotted a few areas that could be improved, but on the whole this is consistent with what I was imagining. Apologies for keeping you waiting.

File: tests/memory_io.rs
Comment (Ralith) [inline tests/memory_io.rs]: This doesn't need a mutable reference.

File: src/output/stream_writer.rs
Comment (cessen) [reply src/output/stream_writer.rs]: Oh, that's weird.  Apparently I left a dead file in.  That's from an earlier iteration, and is dead code.  I realized the same thing, and got rid of that struct entirely.  I'll remove the file.  Sorry!

File: src/output/stream_writer.rs
Comment (cessen) [reply src/output/stream_writer.rs]: Yup!

File: src/stream_io.rs
Comment (cessen) [reply src/stream_io.rs]: I'm happy removing that, yeah.  It also reduces the FFI surface area to take it out, which is nice.  I'll get rid of it.

File: tests/memory_io.rs
Comment (cessen) [reply tests/memory_io.rs:32]: I'm not sure I understand.  It's already zero initialized I thought?  So I'm assuming you mean something different from what I'm understanding zero initialized to mean.

File: tests/memory_io.rs
Comment (cessen) [reply tests/memory_io.rs]: Fair enough.  Stylistically I generally like to try to enclose initialization bits in scopes to make it clear what's related and what's not.  But in this case I think you're right, it doesn't add much clarity.

File: tests/memory_io.rs
Comment (cessen) [reply tests/memory_io.rs]: Oh... that's actually backwards.  It should need to be (the `write_pixels` call modifies the pixel data through the fb reference--effectively, the fb has a mut reference to the pixel data).  I'll fix that in the API.

File: tests/memory_io.rs
Comment (Ralith) [reply tests/memory_io.rs:32]: Oh, wow, I don't know how I misread that. Disregard.

File: tests/memory_io.rs
Comment (Ralith) [reply tests/memory_io.rs]: I think the more conventional idiom for delimiting such logical groups is just empty lines. Saves indentation, at least. Definitely a trivial style quibble, not something there's a Right Answer to and absolutely not a blocker.

File: tests/memory_io.rs
Comment (Ralith) [reply tests/memory_io.rs]: I think there's some confusion there. `write_pixels` is a method on `ScanlineOutputFile`, *reads* pixels from the `FrameBuffer`, and writes them to the *file*, unless I'm very confused myself. The naming's not the best but I think we want to be consistent with upstream here and the types disambiguate it pretty well. The API is correct.

File: openexr-sys/c_wrapper/rust_ostream.cpp
Comment (cessen) [reply openexr-sys/c_wrapper/rust_ostream.cpp]: Ah, thanks for the tip!  Done, to an extent.  It would be cool to figure out a way to turn it back into a proper Rust error (somehow round-trip it through the C++ API) but this seems to work well for now.

File: tests/memory_io.rs
Comment (cessen) [reply tests/memory_io.rs:32]: No worries!

**Comment by cessen** [comment]
Okay, made the changes you suggested (I think... assuming I didn't misunderstand anything).  If you want to give it one more once-over, then we can merge and do the same thing for the input API.

File: src/output/scanline_output_file.rs
Comment (Ralith) [inline src/output/scanline_output_file.rs]: I believe `mut` is not needed here as during output the `FrameBuffer` is only read from, per previous discussion. Contrast the `InputFile::read_pixels` case, where `mut` is needed because the `FrameBuffer` is written to.

File: src/stream_io.rs
Comment (Ralith) [inline src/stream_io.rs]: 1 is not a correct sentinel value here. For example, on Linux, error 1 is EPERM, or "Operation not permitted." Reporting all non-OS `io::Error` values as EPERM would be very confusing. Finding a cross-platform safe sentinel value is probably more difficult than just separating the success flag from the error code.

File: src/stream_io.rs
Comment (Ralith) [inline src/stream_io.rs]: See comment on `write_stream` concerning sentinel value correctness.

File: tests/memory_io.rs
Comment (Ralith) [reply tests/memory_io.rs]: To reiterate, the previous API was correct, having been adjusted from the current erroneous state in previous PRs. Consider that [`OutputFile::setFrameBuffer`](https://github.com/openexr/openexr/blob/develop/OpenEXR/IlmImf/ImfOutputFile.h#L128) takes a const reference, and [`writePixels`](https://github.com/openexr/openexr/blob/develop/OpenEXR/IlmImf/ImfOutputFile.h#L151) certainly gives no reason to believe that OpenEXR makes some wildly unsafe violation of that contract.

**Review by Ralith — CHANGES_REQUESTED** [review]
Still a few correctness issues remaining.

File: openexr-sys/c_wrapper/rust_ostream.cpp
Comment (Ralith) [reply openexr-sys/c_wrapper/rust_ostream.cpp]: Round-tripping back into `io::Error` can be accomplished with a distinct `catch` clause for `std::system_error` in the CEXR bindings, and corresponding expansions of the interface. However, given that throwing `std::system_error` alone should give us good error messages, I don't think that needs to block this PR.

File: src/output/scanline_output_file.rs
Comment (cessen) [reply src/output/scanline_output_file.rs]: Oh!  Yes, indeed.  Good point.  I'll undo that change.

File: src/stream_io.rs
Comment (cessen) [reply src/stream_io.rs]: Ah, fair point.

Comment (cessen) [reply src/stream_io.rs]: Hmm... I do wonder, though.  So, to be clear, your point still stands and I'll address it.  But even so, that still leaves the issues of what to do in case the unwrap fails?  To we statically know that the unwrap will not fail in this case (in which case I should probably use `expect` with a decent message)?

Comment (cessen) [reply src/stream_io.rs]: Okay, so looking into it, we actually don't know statically that it won't fail.  If it's from a `std::fs::File` then the error won't fail, but these functions operate on any type that implements `Write` and `Seek` respectively.

So what I'm thinking is that for now we continue to default to 1, but with a TODO to figure out something better.  And then in the future we can try to figure out a way to fully transparently propagate the Rust `Error` through the C++ API (if that's possible).

For `File` that should still result in the correct behavior once I separate out the success flag and the error code.

Comment (cessen) [reply src/stream_io.rs]: I separated out the error code from the success flag, and also distinguish between system errors and otherwise.  I'm thinking this is good enough for now?

**Comment by cessen** [comment]
Okay, I think I've got everything done now.  Can you look over what I've done with the error handling and see if it makes sense to you?

As I said in some other comments, ideally I'd like to find a way to transparently pass the rust error through the C++ code.  But for now I'm thinking this is good enough.

**Review by Ralith — APPROVED** [review]
Looks good to me! Probably want to squash before merging.

**Comment by cessen** [comment]
Yay!  Okay, I'm going to do the same thing for input now.

Thanks for your help!

**Pull request merged**
